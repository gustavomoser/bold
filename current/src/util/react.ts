import { MutableRefObject, Ref } from 'react'

/**
 * Compose multiple react refs.
 *
 * @returns A new ref that passes down the Element to all ref arguments.
 */
export function composeRefs<T>(...refs: Array<Ref<T>>) {
  return (value: T) => refs.forEach(ref => setRef(ref, value))
}

/**
 * Changes the value of a ref.
 * The ref can be a function ref or a object ref.
 *
 * @param ref The ref to have its value changed.
 * @param value The new value.
 */
export function setRef<T>(ref: Ref<T>, value: any) {
  if (typeof ref === 'function') {
    ref(value)
  } else if (ref) {
    const mutableRef = ref as MutableRefObject<T>
    mutableRef.current = value
  }
}

/**
 * Compose multiple event handlers.
 *
 * @returns A new function that triggers all handlers passed as parameters, in order.
 */
export function composeHandlers(...handlers: Array<(...params: any) => void>) {
  return (...args: any[]) => handlers.forEach(handler => handler && handler(...args))
}

/**
 * Invokes the native `value` property setter of an element.
 * From https://github.com/facebook/react/issues/10135#issuecomment-401496776
 *
 * @param element The element to invoke the setter on.
 * @param value The value to be set.
 */
export function setNativeValue(element: HTMLElement, value: any) {
  const prototype = Object.getPrototypeOf(element)
  const { set: prototypeValueSetter = null } = Object.getOwnPropertyDescriptor(prototype, 'value') || {}
  const { set: valueSetter = null } = Object.getOwnPropertyDescriptor(element, 'value') || {}

  if (prototypeValueSetter && valueSetter !== prototypeValueSetter) {
    prototypeValueSetter.call(element, value)
  } else if (valueSetter) {
    valueSetter.call(element, value)
  } else {
    throw new Error('The given element does not have a value setter')
  }
}
