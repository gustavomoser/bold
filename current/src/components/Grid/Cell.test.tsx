import { render } from '@testing-library/react'
import React from 'react'

import { Cell } from './Cell'

it('should render correctly', () => {
  const { container } = render(<Cell>Content</Cell>)
  expect(container).toMatchSnapshot()
})

it('should accept flex item props', () => {
  const { container } = render(
    <Cell size={6} alignSelf='center' flexGrow={1} flexShrink={1} flexBasis='50%'>
      Content
    </Cell>
  )
  expect(container).toMatchSnapshot()
})

it('should accept "style" prop', () => {
  const { container } = render(<Cell style={{ color: 'red' }}>Content</Cell>)
  expect(container).toMatchSnapshot()
})

it('should pass props down to div', () => {
  const { container } = render(
    <Cell id='div' aria-label='test'>
      Content
    </Cell>
  )
  expect(container.querySelector('div').getAttribute('id')).toEqual('div')
  expect(container.querySelector('div').getAttribute('aria-label')).toEqual('test')
})

it('should accept size', () => {
  const { container } = render(<Cell size={6} />)
  expect(container).toMatchSnapshot()
})

it('should accept responsive sizes', () => {
  const { container } = render(<Cell xs={12} sm={6} md={4} lg={3} xl={2} />)
  expect(container).toMatchSnapshot()
})
