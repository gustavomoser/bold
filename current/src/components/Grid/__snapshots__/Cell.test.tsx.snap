// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should accept "style" prop 1`] = `
.emotion-0 {
  padding-left: 1rem;
  padding-right: 1rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  color: red;
}

<div>
  <div
    class="emotion-0"
  >
    Content
  </div>
</div>
`;

exports[`should accept flex item props 1`] = `
.emotion-0 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-flex-basis: 50%;
  -ms-flex-preferred-size: 50%;
  flex-basis: 50%;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  padding-left: 1rem;
  padding-right: 1rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  width: calc(100% / 12 * 6);
}

<div>
  <div
    class="emotion-0"
  >
    Content
  </div>
</div>
`;

exports[`should accept responsive sizes 1`] = `
.emotion-0 {
  padding-left: 1rem;
  padding-right: 1rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  width: calc(100% / 12 * 12);
}

@media (min-width:600px) {
  .emotion-0 {
    width: calc(100% / 12 * 6);
  }
}

@media (min-width:960px) {
  .emotion-0 {
    width: calc(100% / 12 * 4);
  }
}

@media (min-width:1280px) {
  .emotion-0 {
    width: calc(100% / 12 * 3);
  }
}

@media (min-width:1920px) {
  .emotion-0 {
    width: calc(100% / 12 * 2);
  }
}

<div>
  <div
    class="emotion-0"
  />
</div>
`;

exports[`should accept size 1`] = `
.emotion-0 {
  padding-left: 1rem;
  padding-right: 1rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  width: calc(100% / 12 * 6);
}

<div>
  <div
    class="emotion-0"
  />
</div>
`;

exports[`should render correctly 1`] = `
.emotion-0 {
  padding-left: 1rem;
  padding-right: 1rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
}

<div>
  <div
    class="emotion-0"
  >
    Content
  </div>
</div>
`;
